src/vehicle.cpp

```diff
@@ -1156,29 +1144,43 @@ void ViewportAddVehicles(DrawPixelInfo *dpi)
 			while (v != nullptr) {
 
 				if (!(v->vehstatus & VS_HIDDEN) &&
-						l <= v->coord.right &&
-						t <= v->coord.bottom &&
-						r >= v->coord.left &&
-						b >= v->coord.top) {
-					DoDrawVehicle(v);
-				}
-				else if (l <= v->coord.right + xb &&
+					l <= v->coord.right + xb &&
 					t <= v->coord.bottom + yb &&
 					r >= v->coord.left - xb &&
 					b >= v->coord.top - yb)
 				{
 					/*
-					 * Indicate that this vehicle was considered for rendering in a viewport,
-					 * is within the bounds where a sprite could be valid for rendering
-					 * and we therefore need to update sprites more frequently in case a callback
-					 * will change the bounding box to one which will cause the sprite to be
-					 * displayed.
-					 *
-					 * This reduces the chances of flicker when sprites enter the screen, if they
-					 * are part of a newgrf vehicle set which changes bounding boxes within a
-					 * single vehicle direction.
+					 * This vehicle can potentially be drawn as part of this viewport and
+					 * needs to be revalidated, as the sprite may not be correct.
 					 */
-					v->sprite_cache.is_viewport_candidate = true;
+					if (v->sprite_cache.revalidate_before_draw) {
+						VehicleSpriteSeq seq;
+						v->GetImage(v->direction, EIT_ON_MAP, &seq);
+
+						if (v->sprite_cache.sprite_seq != seq) {
+							v->sprite_cache.sprite_seq = seq;
+
+							/*
+							 * A sprite change may also result in a bounding box change,
+							 * so we need to update the bounding box again before we
+							 * check to see if the vehicle should be drawn. Note that
+							 * we can't interfere with the viewport hash at this point,
+							 * so we keep the original hash on the assumption there will
+							 * not be a significant change in the top and left coordinates
+							 * of the vehicle.
+							 */
+							Vehicle* v_mutable = const_cast<Vehicle*>(v);
+							v_mutable->Vehicle::UpdateBoundingBoxCoordinates(false);
```

Comment:
Bit irritating that the const_cast has had to be reintroduced here. Removing it was the whole point of the MutableSpriteCache structure :)

Reply:
I'll have a short play around with this, see if there's a fix which doesn't result in wholesale code changes.

Reply:
Latest change moves this chain to const, with vehicle coords becoming mutable.